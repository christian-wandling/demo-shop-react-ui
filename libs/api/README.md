# api

A comprehensive library for consuming the demo-shop api. Supports accessing a product catalog, user management,
shopping cart operations, and order processing.
