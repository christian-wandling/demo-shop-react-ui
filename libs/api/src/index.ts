export * from './lib/openapi';
export * from './lib/ApiContext';
export * from './lib/services/apiService';
export * from './lib/ApiProvider';
export * from './lib/services/apiService';
export * from './lib/middleware/authMiddleware';
