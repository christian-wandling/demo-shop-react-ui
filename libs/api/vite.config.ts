import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig(() => ({
  root: __dirname,
  cacheDir: '../../node_modules/.vite/libs/api',
  plugins: [react()],
  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },
  test: {
    watch: false,
    globals: true,
    environment: 'jsdom',
    reporters: ['default'],
    passWithNoTests: true,
    coverage: {
      reportsDirectory: './test-output/vitest/coverage',
      enabled: false,
    },
  },
}));
