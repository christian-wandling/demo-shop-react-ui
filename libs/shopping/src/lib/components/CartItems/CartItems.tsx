import { Link } from 'react-router';
import { CartItemsProps } from '../../models/cartItemsProps';

/**
 * CartItems displays items in a shopping cart.
 *
 * This component is designed to render cart items and provide functionality
 * to interact with them, such as removing items from the cart.
 *
 * @example
 * <CartItems items={cartItems} onRemoveItem={handleRemoveItem} />
 */
export const CartItems = ({ items, onRemoveItem }: CartItemsProps) => {
  return (
    <div className="mt-8">
      <div className="flow-root">
        <ul className="-my-6 divide-y divide-gray-200">
          {items.map(item => (
            <li key={item.id} className="flex py-6">
              <Link
                to={`/products/${item.productId}`}
                className="relative h-24 w-24 aspect-1 flex-shrink-0 overflow-hidden rounded-md border border-gray-200 group-hover:opacity-75 cursor-pointer">
                <img
                  src={item.productThumbnail}
                  alt={item.productName}
                  title={item.productName}
                  className="object-cover object-center"
                />
              </Link>

              <div className="ml-4 flex flex-1 flex-col">
                <div>
                  <div className="flex justify-between text-base font-medium text-gray-900">
                    <h3>
                      <a>{item.productName}</a>
                    </h3>
                    <p className="ml-4">${item.totalPrice.toFixed(2)}</p>
                  </div>
                </div>
                <div className="flex flex-1 items-end justify-between text-sm">
                  <p className="text-gray-500">Qty {item.quantity}</p>

                  <div className="flex">
                    <button
                      onClick={() => onRemoveItem(item.id)}
                      type="button"
                      className="font-medium text-indigo-600 hover:text-indigo-500">
                      Remove
                    </button>
                  </div>
                </div>
              </div>
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};
