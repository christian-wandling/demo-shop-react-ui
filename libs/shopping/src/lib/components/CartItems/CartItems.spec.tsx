import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { CartItems } from './CartItems';
import { BrowserRouter } from 'react-router';
import { CartItemResponse } from '@demo-shop-react-ui/api';

const mockCartItems: CartItemResponse[] = [
  {
    id: 1,
    productId: 101,
    productName: 'Test Product 1',
    productThumbnail: 'https://example.com/image1.jpg',
    quantity: 2,
    totalPrice: 19.99,
    unitPrice: 0,
  },
  {
    id: 2,
    productId: 102,
    productName: 'Test Product 2',
    productThumbnail: 'https://example.com/image2.jpg',
    quantity: 1,
    totalPrice: 29.99,
    unitPrice: 0,
  },
];

const renderWithRouter = (component: any) => {
  return render(<BrowserRouter>{component}</BrowserRouter>);
};

describe('CartItems Component', () => {
  it('renders successfully', () => {
    const onRemoveItem = vi.fn();
    renderWithRouter(<CartItems items={mockCartItems} onRemoveItem={onRemoveItem} />);
    expect(screen.getByText('Test Product 1')).toBeTruthy();
  });

  it('matches the snapshot', () => {
    const onRemoveItem = vi.fn();
    const { baseElement } = renderWithRouter(<CartItems items={mockCartItems} onRemoveItem={onRemoveItem} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('renders all items in the cart', () => {
    const onRemoveItem = vi.fn();
    renderWithRouter(<CartItems items={mockCartItems} onRemoveItem={onRemoveItem} />);

    expect(screen.getByText('Test Product 1')).toBeTruthy();
    expect(screen.getByText('Test Product 2')).toBeTruthy();

    expect(screen.getByText('Qty 2')).toBeTruthy();
    expect(screen.getByText('Qty 1')).toBeTruthy();

    expect(screen.getByText('$19.99')).toBeTruthy();
    expect(screen.getByText('$29.99')).toBeTruthy();
  });

  it('calls onRemoveItem with correct item id when Remove button is clicked', () => {
    const onRemoveItem = vi.fn();
    renderWithRouter(<CartItems items={mockCartItems} onRemoveItem={onRemoveItem} />);

    const removeButtons = screen.getAllByText('Remove');
    fireEvent.click(removeButtons[0]);

    expect(onRemoveItem).toHaveBeenCalledTimes(1);
    expect(onRemoveItem).toHaveBeenCalledWith(1);

    fireEvent.click(removeButtons[1]);
    expect(onRemoveItem).toHaveBeenCalledTimes(2);
    expect(onRemoveItem).toHaveBeenLastCalledWith(2);
  });

  it('renders product images with correct attributes', () => {
    const onRemoveItem = vi.fn();
    renderWithRouter(<CartItems items={mockCartItems} onRemoveItem={onRemoveItem} />);

    const images: HTMLImageElement[] = screen.getAllByRole('img');
    expect(images).toHaveLength(2);

    expect(images[0].src).toBe('https://example.com/image1.jpg');
    expect(images[0].alt).toBe('Test Product 1');
    expect(images[0].title).toBe('Test Product 1');

    expect(images[1].src).toBe('https://example.com/image2.jpg');
    expect(images[1].alt).toBe('Test Product 2');
    expect(images[1].title).toBe('Test Product 2');
  });

  it('renders product links with correct routes', () => {
    const onRemoveItem = vi.fn();
    renderWithRouter(<CartItems items={mockCartItems} onRemoveItem={onRemoveItem} />);

    const productLinks: NodeListOf<HTMLLinkElement> = document.querySelectorAll('a[href^="/products/"]');
    expect(productLinks).toHaveLength(2);

    expect(productLinks[0].href).toContain('/products/101');
    expect(productLinks[1].href).toContain('/products/102');
  });

  it('renders empty list when no items are provided', () => {
    const onRemoveItem = vi.fn();
    renderWithRouter(<CartItems items={[]} onRemoveItem={onRemoveItem} />);

    const listItems = screen.queryAllByRole('listitem');
    expect(listItems).toHaveLength(0);
  });
});
