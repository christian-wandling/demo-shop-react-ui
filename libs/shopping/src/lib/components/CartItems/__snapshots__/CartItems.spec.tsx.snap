// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartItems Component > matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="mt-8"
    >
      <div
        class="flow-root"
      >
        <ul
          class="-my-6 divide-y divide-gray-200"
        >
          <li
            class="flex py-6"
          >
            <a
              class="relative h-24 w-24 aspect-1 flex-shrink-0 overflow-hidden rounded-md border border-gray-200 group-hover:opacity-75 cursor-pointer"
              data-discover="true"
              href="/products/101"
            >
              <img
                alt="Test Product 1"
                class="object-cover object-center"
                src="https://example.com/image1.jpg"
                title="Test Product 1"
              />
            </a>
            <div
              class="ml-4 flex flex-1 flex-col"
            >
              <div>
                <div
                  class="flex justify-between text-base font-medium text-gray-900"
                >
                  <h3>
                    <a>
                      Test Product 1
                    </a>
                  </h3>
                  <p
                    class="ml-4"
                  >
                    $
                    19.99
                  </p>
                </div>
              </div>
              <div
                class="flex flex-1 items-end justify-between text-sm"
              >
                <p
                  class="text-gray-500"
                >
                  Qty 
                  2
                </p>
                <div
                  class="flex"
                >
                  <button
                    class="font-medium text-indigo-600 hover:text-indigo-500"
                    type="button"
                  >
                    Remove
                  </button>
                </div>
              </div>
            </div>
          </li>
          <li
            class="flex py-6"
          >
            <a
              class="relative h-24 w-24 aspect-1 flex-shrink-0 overflow-hidden rounded-md border border-gray-200 group-hover:opacity-75 cursor-pointer"
              data-discover="true"
              href="/products/102"
            >
              <img
                alt="Test Product 2"
                class="object-cover object-center"
                src="https://example.com/image2.jpg"
                title="Test Product 2"
              />
            </a>
            <div
              class="ml-4 flex flex-1 flex-col"
            >
              <div>
                <div
                  class="flex justify-between text-base font-medium text-gray-900"
                >
                  <h3>
                    <a>
                      Test Product 2
                    </a>
                  </h3>
                  <p
                    class="ml-4"
                  >
                    $
                    29.99
                  </p>
                </div>
              </div>
              <div
                class="flex flex-1 items-end justify-between text-sm"
              >
                <p
                  class="text-gray-500"
                >
                  Qty 
                  1
                </p>
                <div
                  class="flex"
                >
                  <button
                    class="font-medium text-indigo-600 hover:text-indigo-500"
                    type="button"
                  >
                    Remove
                  </button>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</body>
`;
